### PR TITLE
SLING-11427 : Registering a Namespace does not allow Hashtag in the URI

### PR DESCRIPTION
**Pull request by schaefa**
I am no Javacc expert but that seems to do the trick.

I allow for an escaped hashtag inside a string and then replace \# with a # inside the RegisterNamespace operation.

### CODE REVIEW

**Comment by bdelacretaz** [comment]
We already have a `quotableString()` production in the grammar, would it be an option to allow for quoting the namespace URI for such cases? The input would then be

    register namespace (foo2) "uri:some-uri/V/1.1/test#"

**Comment by enapps-enorman** [comment]
I think I'd agree with @bdelacretaz that the quotableString() would probably cover more use cases.  I believe there are additional characters that are allowed in a namespace URI that are not matched by the RepoInitGrammar's <STRING> token. For example: [~, ?, &, %, =] and likely many other characters.

Some invalid values may be matched by quotableString() as well, so it may be prudent to do additional validation within the RegisterNamespace operation to make sure the string can be parsed as a valid JCR namespace URI.

**Comment by schaefa** [comment]
I tried that but I could not get the '#' in. Either it is ignored or I have '\#'. The quotedString() is not converting an escaped backslash. Even when I fix the conversion in the RegisterNamespace operation class I get a RepIniParsingException because it does not like the '\#' or '\\#'.
Any suggestions?

**Comment by schaefa** [comment]
I updated the code to showcase the issue. I tried multiple things:
\#: fails the repo init parser test
\\#: fails the repo init parser test
\\\\#: fails the output and repo init parser test

File: src/test/resources/testcases/test-40-output.txt [hunk 1]
```diff
@@ -1,3 +1,3 @@
 RegisterNamespace (foo) uri:some-uri/V/1.0
 RegisterNamespace (prefix_with-other.things) andSimpleURI
-RegisterNamespace (foo2) uri:some-uri/V/1.1/test#
+RegisterNamespace (foo2) uri:some-uri/V/1.1/test#
```
Comment (enapps-enorman) [inline src/test/resources/testcases/test-40-output.txt:3]: The URI should be wrapped with quotes here in the output

File: src/test/resources/testcases/test-40.txt [hunk 2]
```diff
@@ -3,4 +3,4 @@
 # and o.a.s.jcr.repoinit 1.0.2
 register namespace (foo) uri:some-uri/V/1.0
 register namespace ( prefix_with-other.things ) andSimpleURI
-register namespace (foo2) uri:some-uri/V/1.1/test\#
\ No newline at end of file
+register namespace (foo2) "uri:some-uri/V/1.1/test\\#"
```
Comment (enapps-enorman) [inline src/test/resources/testcases/test-40.txt:6]: Should not need to escape the hash character here since the quotableString() doesn't need it.

**Review by enapps-enorman — CHANGES_REQUESTED** [review]
I believe you will also need to adjust the RegisterNamespace#asRepoInitString code to handle the uri field being quotable.  For example the value should call something like: `QuotableStringUtil.forRepoInitString(uri)`

Though, it looks to me like the QuotableStringUtil#REQUIRES_QUOTES regex pattern will be insufficient to detect that the hash character needs quotes.  So perhaps that regex (and logic) should be changed to match what doesn't require quotes and put quotes around everything else?

**Comment by schaefa** [comment]
I did that but now the test fails because the hashtag went missing

**Comment by enapps-enorman** [comment]
> I did that but now the test fails because the hashtag went missing

@schaefa  Take a look at the what I stashed in my fork at: https://github.com/enapps-enorman/sling-org-apache-sling-repoinit-parser/commit/6d8a5401468d4965cc1726236dd7dac206df7cbc to see if that works for you.

**Comment by schaefa** [comment]
@enapps-enorman I used your suggestions but that requires the output to have quotes when it contains special character like the hashtag (RegisterNamespace (foo2) "uri:some-uri/V/1.1/test#") - I hope that works

**Comment by enapps-enorman** [comment]
> @enapps-enorman I used your suggestions but that requires the output to have quotes when it contains special character like the hashtag (RegisterNamespace (foo2) "uri:some-uri/V/1.1/test#") - I hope that works

I think that is because you have a change in RegisterNamespace#getParametersDescription that I did not have.  To be honest, I'm not sure if quotes or no-quotes is correct in the "toString" output.  It looked to me that other places were not using quotes in the "toString" output for their quotable strings.

**Comment by schaefa** [comment]
Yes, that fixed the issue

**Review by enapps-enorman — APPROVED** [review]

**Pull request merged**